# 정규 표현식(Regular Expression, Regex) 개요

## 정규 표현식이란
- 정규표현식은 특정한 패턴을 가진 문자열을 검색, 추출, 또는 수정하기 위해 사용되는 기법이다.
- 파이썬 뿐만 아니라 문자열을 다루는 모든 곳에서 사용된다.

## 주요 용도

- **문자열 검색**: 특정 패턴에 맞는 문자열을 찾는다.
- **문자열 변경**: 패턴에 맞는 부분 문자열을 다른 문자열로 변경하거나 삭제한다.
- **문자열 유효성 검사**: 이메일 주소, 전화번호와 같이 특정 형식을 확인하여 유효성을 검사합니다.

## 정규표현식 구성 요소
- **패턴** 
    - 찾으려는 문자열의 규칙을 정의한 표현식을 패턴이라고 한다.
    - 정규표현식 패턴은 메타문자와 리터럴로 구성된다.
- **메타문자**
    - 패턴에서 특정 규칙이나 조건을 기술하기 위해 사용되는 특별한 의미를 가지는 문자.
    - 예) `a*` : a가 0회 이상 반복을 뜻한다. a, aa, aaaa
- **리터럴**
    - 문자나 문자열을 패턴 내에서 그 자체로 사용하는 것을 말한다. 
    - 예) `a`는 `a` 자체를 의미한다.    

# 정규 표현식 메타 문자

## 문자 클래스 :  [  ]
- `[ ]` 사이의 문자들과 매칭
    - `[abc]` : a, b, c 중 **하나의 문자**와 매치
- `-`를 이용해 범위로 설정할 수 있다.
    - `[a-z]` : 알파벳소문자중 하나의 문자와 매치
    - `[a-zA-Z0-9]` : 알파벳대소문자와 숫자 중 하나의 문자와 매치
    - `[가-힣ㄱ-ㅎㅏ-ㅣ]`: 한글중 하나와 매치
- `[^ 패턴]` : ^ 으로 시작하는 경우 반대의 의미. 와서 안되는 패턴을 의미
    - `[^abc]` : a, b, c를 제외한 나머지 문자들 중 하나와 매치.
    - `[^a-z]` : 알파벳 소문자를 제외한 나머지 문자들 중 하나와 매치

## 미리 정의된 문자 클래스
- 자주 사용되는 문자클래스를 미리 정의된 별도 표기법으로 제공한다.
- `\d` : 숫자와 매치. [0-9]와 동일
- `\D` : `\d`의 반대. 숫자가 아닌 문자와 매치.  [^0-9]와 동일
- `\w` : 문자와 숫자, _(underscore)와 매치. `[a-zA-Z가-힣0-9_]`와 동일  (문자는 특수문자 제외한 일반문자-언어상관없는-들을 말한다.
- `\W` : `\w`의 반대. 문자와 숫자와 _ 가 아닌 문자와 매치.  `[^a-zA-Z가-힣0-9_]`와 동일
- `\s` : 공백문자와 매치. tab,줄바꿈,공백문자와 일치
- `\S` : `\s`와 반대. 공백을 제외한 문자열과 매치.
- `\b` : 단어 경계(word boundary) 표시. \b가 매칭하는 경계문자로 공백, `,`, `.`, `\n` 이 있다.
    - \b는 문자 자체와 매칭하지 않고 경계만 확인한다. 단어의 시작과 단어의 끝을 찾는 데 유용하다.
    - `\bcat\b`는 "cat"이라는 단어가 정확히 존재하는 경우만 매칭한다. 그래서 "**cat**egory"나 "con**cat**enate" 같은 단어에는 매칭되지 않고 "The cat is" 에서 cat 에만 매칭된다.
- `\B` : `\b`의 반대. 단어 경계로 구분된 단어가 아닌 경우
    - `\B가족\B` => 우리 가족 만세(X), 우리가족만세 (O)

## 수량자(quantifier, 글자수와 관련된 메타문자)
- `*` : 앞의 문자(패턴)과 일치하는 문자가 0개 이상인 경우. (`a*b`)
- `+` : 앞의 문자(패턴)과 일치하는 문자가 1개이상인 경우.  (`a+b`)
- `?` :  앞의 문자(패턴)과 일치하는 문자가 한개 있거나 없는 경우. (`a?b`)
- `{m}` : 앞의 문자(패턴)가 m개. (`a{3}b`)
- `{m,}` : 앞의 문자(패턴)이 m개 이상. (`a{3,}b`)
    - , 뒤에 공백이 들어오지 않도록 한다.
- `{m,n}` : 앞의 문자(패턴)이 m개이상 n개 이하. (`a{2,5}b`)

## 문장의 시작과 끝 표현
- `^` 문자열의 시작 (`^abc`)
    - 문자 클래스([ ])의 ^와는 의미가 다르다.
- `$` : 문자열의 끝 (`abc$`)

## 기타
- `.` : 한개의 모든 문자(\n-줄바꿈 제외) (`a.b`)
- `|` : 둘중 하나(OR) (?:010|011|016|019)
    - 010|016-111 : 010 또는 016-111 이 된다. 
- `(  )` : 패턴내 하위그룹을 만들때 사용

# regexr.com

- 정규표현식을 테스트하고 검증할 수있는 사이트
- https://regexr.com/

# 파이썬에서 정규표현식 사용하기
- 표준 모듈 `re` 를 사용한다.
    - re는 정규표현식을 전용 모듈이고 다양한 패키지들이 내부적으로 정규표현식을 사용한다.
- 파이썬에서 정규 표현식을 지원하기 위한 모듈
- 파이썬 기본 라이브러리

## 코딩 방식

1. 객체지향형
    - 패턴 객체를 생성후 메소드를 호출해 원하는 처리를 한다.
     ```python
        p = re.compile(r'\d+')
        p.search('abc123def')
    ```
2. 함수형
    - `re` 모듈의 원하는 작업을 하는 함수를 호출한다. Argument로 패턴과 처리할 값을 전달한다.
    ```python
        re.search(r'\d+', 'abc123def')
    ```
    
> ### raw string
> - 파이썬은 문자열에 `\` 가 있으면 우선적으로 escape 문자로 처리한다. 그런데 메타문자 중 "미리 정의된 문자클래스" 들은 다 `\`로 시작한다.  그래서 이들을 사용할 경우 `escape` 문자와의 구분을 위해 `\\` 두개씩 작성해야한다.  이를 피하기 위해 패턴을 작성할 때는 raw string을 사용하는 것이 편리하다.
>    - `re.compile('\b가족\b')` : `\b`를 escape 문자 b(백스페이스)로 인식
>    - `re.compile(r'\b가족\b')` : `\b`가 일반문자가 되어 컴파일시 정규식 메타문자로 처리된다.


## 검색함수
- match(), search() : 패턴과 일치하는 문장이 **있는지 여부**를 확인할 때 사용
- findall(), finditer(s) : 패턴과 일치하는 문장을 **찾을 때** 사용

### Match class
- **검색 결과 하나를** 담는 class
    - match(), search() 의 반환타입으로 검색결과를 담는다.
- 패턴과 일치한 문자열과 그 문자열의 위치를 가진다.
- 주요 메소드
    - **group()** : 찾은 결과(문자열)을 반환
    - **group(subgroup 번호)** : 패턴에 하위그룹이 지정된 경우 특정 그룹의 문자열 반환
    - **start(), end()** : 대상 문자열내에서 시작, 끝 index 반환
    - **span()** : 대상 문자열 내에서 시작, 끝 index를 tuple로 반환

### match(대상문자열 [, pos=0])
- 대상 문자열의 시작이 정규식과 일치하는지를 조회.
- pos : 시작 index 지정
- 반환값
    - Match 객체: 일치하는 문자열이 있는 경우
    - None: 일치하는 문자열이 없는 경우

### search(대상문자열 [, pos=0])
- 대상문자열 전체 안에서 정규식과 일치하는 것이 있는지 조회
- pos: 찾기 시작하는 index 지정
- 반환값
    - Match 객체: 일치하는 문자열이 있는 경우
    - None: 일치하는 문자열이 없는 경우|

In [1]:
txt1 = "안녕하세요. 반갑습니다. 오늘 날씨는 3도입니다. 좋은 하루 되세요."
txt2 = "반갑습니다. 안녕하세요. 오늘 날씨는 3도입니다. 좋은 하루 되세요. 반갑습니다."
txt3 = "가격은 각각 1000, 2000, 3000, 60000원 입니다."

In [2]:
import re

In [15]:
# match() - 객체지향방식
p = re.compile(r"[가-힣]+습니다\.") # [가-힣] 모든 한글문자 1글자.
                           # + 수량사로 앞의 패턴이 한글자 이상
                           # 습니다 -> literal(리터럴)
                           # `\.`  .: 모든 한글자. 마침표: \. 
                           # ~~~~~습니다.
# result = p.match(txt1) # txt1이 p의 패턴으로 **시작하는지** 여부.
# result = p.match(txt2)
# result = p.match(txt2, pos=7) # index 7 에서부터 찾는다.

# result = p.search(txt1)  # txt1안에 패턴의 문자열이 있는지 검색.
# result = p.search(txt2)
result = p.search(txt2, pos=5) # index 5에서부터 찾는다.

if result != None:
    print(result)
    print(result.group()) # 찾은 문자열을 반환.
    print(result.start(), result.end()) # 찾은 문자열의 시작/끝-1 위치
    print(result.span())
else: # None
    print(result, "패턴이 일치하지 않습니다.")

<re.Match object; span=(39, 45), match='반갑습니다.'>
반갑습니다.
39 45
(39, 45)


In [26]:
txt3 = "가격은 각각 800000, 1000, 2000, 3000, 60000 원 입니다."

# 함수형 방식
import re
# txt3에서 만단위 숫자가 있는지 확인
result = re.search(r"\b\d{5}\b", txt3) # , pos=3) [0-9] -> \d, {5} 5개
print(result)

# 객체지향방식
p = re.compile(r"\b\d{5}\b")
print(p.search(txt3))
print(p.match(txt3))



<re.Match object; span=(33, 38), match='60000'>
<re.Match object; span=(33, 38), match='60000'>
None


### findall(대상문자열)
- 대상문자열에서 정규식과 매칭되는 문자열들을 리스트로 반환
- 반환값
    - 리스트(List) : 일치하는 문자열들을 가진 리스트를 반환
    - 일치하는 패턴이 없을 경우 빈 리스트를 반환한다.
    
### finditer(대상문자열)
- 대상문자열에서 정규식과 매칭되는 결과들을 조회할 수있는 Iterator를 반환한다.
- 반환값
    - callable_iterator
    - 일치하는 패턴이 없어도 iterator객체는 반환되는데 next()시 StopIteration Exception발생한다.

In [30]:
# txt3에서 숫자들을 모두 조회
p = re.compile(r"\d+")  

result = p.findall(txt3)
result

['800000', '1000', '2000', '3000', '60000']

In [40]:
result = p.finditer(txt3)
result

In [35]:
# list(result)

In [41]:
for r in result:
    print(r.group(), r.span())

800000 (7, 13)
1000 (15, 19)
2000 (21, 25)
3000 (27, 31)
60000 (33, 38)


### TODO
- info 변수는 한줄에 한사람의 data가 있고 구성은 **`이름 이메일주소 주민번호`** 순서로 되어있다.

In [45]:
info ='''김정수 kjs@gmail.com 801023-1010221
박영수 pys.abc@gmail.com 700121-1120212
이민영 lmy-abc@naver.com 820301-2020122
김순희 ksh@daum.net 781223-2012212
오주연 ojy@daum.net 900522-1023218
'''

In [47]:
# 주민번호들만 조회해서 출력
# 숫자6-숫자7
p = re.compile(r"\d{6}-[1-8]\d{6}")
result = p.findall(info)
print(result)
result2 = p.finditer(info)
list(result2)

['801023-1010221', '700121-1120212', '820301-2020122', '781223-2012212', '900522-1023218']


[<re.Match object; span=(18, 32), match='801023-1010221'>,
 <re.Match object; span=(55, 69), match='700121-1120212'>,
 <re.Match object; span=(92, 106), match='820301-2020122'>,
 <re.Match object; span=(124, 138), match='781223-2012212'>,
 <re.Match object; span=(156, 170), match='900522-1023218'>]

In [57]:
import string

# list(string.punctuation)
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [50]:
# Email 주소만 추출 해서 출력 - 계정@도메인
# p = re.compile(r"[a-z0-9A-Z-.!#]+@[a-z0-9A-Z-.!#]+")
p = re.compile(r"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])")
result = p.findall(info)
print(result)

['kjs@gmail.com', 'pys.abc@gmail.com', 'lmy-abc@naver.com', 'ksh@daum.net', 'ojy@daum.net']


In [59]:
result = p.finditer(info)
for value in result:
    print(value)

<re.Match object; span=(4, 17), match='kjs@gmail.com'>
<re.Match object; span=(37, 54), match='pys.abc@gmail.com'>
<re.Match object; span=(74, 91), match='lmy-abc@naver.com'>
<re.Match object; span=(111, 123), match='ksh@daum.net'>
<re.Match object; span=(143, 155), match='ojy@daum.net'>


## 문자열 변경
- sub(): 변경된 문자열 반환
- subn(): 변경된 문자열, 변경개수 반환

### sub(바꿀문자열, 대상문자열 [, count=양수])
- 대상문자열에서 패턴과 일치하는 것을 바꿀문자열로 변경한다.
- count: 변경할 개수를 지정. 기본: 매칭되는 문자열은 다 변경
- 반환값: 변경된 문자열

### subn(바꿀문자열, 대상문자열 [, count=양수])
- sub()와 동일한 역할.
- 반환값 : (변경된 문자열, 변경된문자열개수) 를 tuple로 반환

In [64]:
txt = "   오늘은       목요일            입니다.                     "

p = re.compile(r" +")  # r"a+" " +" 공백 1개 이상.   \s+ 공백문자(" ", \n, \t)
result = p.sub(" ", txt)
print(txt)
print(result.strip())

   오늘은       목요일            입니다.                     
오늘은 목요일 입니다.


In [63]:
result2 = p.subn(" ", txt)
print(result2)

(' 오늘은 목요일 입니다. ', 4)


In [66]:
result3 = re.sub(r" +", " ", txt).strip()
print(result3)

오늘은 목요일 입니다.


# Grouping - 패턴내 하위 패턴 만들기
- 패턴의 일부를 하나의 그룹으로 묶는 기능으로, 매칭된 패턴의 일부를 재사용하거나, 특정 패턴이 일치하는지 확인할 때 유용.
    - 보통 패턴이 여러개의 하위 패턴(속성)들로 구성되 있고 전체 내용에서 일부 속성들을 매칭 시켜야 할 때 사용한다.
    - 예를 들어, 전화번호는 "지역번호/010-국번-번호" 형식으로 구성된다. 이때 패턴을 만들면서 국번 부분을 그룹화하면, 전화번호를 찾은 후 국번만 쉽게 추출하여 조회할 수 있다.

- 구문: 하위 패턴을 **소괄호**로 묶어준다.
    - `(\d{4})/([01]\d)/([0123]\d)`  
    - (년도)/(월)/(일)

## 그룹핑 예

### 전체 패턴과 매칭된 결과에서 하위 패턴을 조회

In [67]:
txt = """tel)010-2323-3232
tel)02-323-9999
tel)031-233-4841
tel)010-1112-3344"""
print(txt)

tel)010-2323-3232
tel)02-323-9999
tel)031-233-4841
tel)010-1112-3344


In [75]:
p = re.compile(r"0\d{1,2}-\d{3,4}-\d{4}") # 0\d{1,2} : 0다음에 숫자1 또는 2개
result = p.finditer(txt)

for tel in result:
    print(tel, tel.group(), tel.group().split("-")[1], sep=", ")

<re.Match object; span=(4, 17), match='010-2323-3232'>, 010-2323-3232, 2323
<re.Match object; span=(22, 33), match='02-323-9999'>, 02-323-9999, 323
<re.Match object; span=(38, 50), match='031-233-4841'>, 031-233-4841, 233
<re.Match object; span=(55, 68), match='010-1112-3344'>, 010-1112-3344, 1112


In [81]:
# grouping을 이용해서 국번만 조회
# 패턴 "( 1 )-( 2 )-( 3 )"
p = re.compile(r"(0\d{1,2})-(\d{3,4})-(\d{4})") # 0\d{1,2} : 0다음에 숫자1 또는 2개
result = p.finditer(txt)
for value in result:
    print("전체 text:", value.group(), value.group(0))
    print("1번 하위그룹:", value.group(1))
    print("2번 하위그룹:", value.group(2))
    print("3번 하위그룹:", value.group(3))
    print("--------------------")

전체 text: 010-2323-3232 010-2323-3232
1번 하위그룹: 010
2번 하위그룹: 2323
3번 하위그룹: 3232
--------------------
전체 text: 02-323-9999 02-323-9999
1번 하위그룹: 02
2번 하위그룹: 323
3번 하위그룹: 9999
--------------------
전체 text: 031-233-4841 031-233-4841
1번 하위그룹: 031
2번 하위그룹: 233
3번 하위그룹: 4841
--------------------
전체 text: 010-1112-3344 010-1112-3344
1번 하위그룹: 010
2번 하위그룹: 1112
3번 하위그룹: 3344
--------------------


In [86]:
# ( 1 )-(2 ( 3 )-( 4 ))
p = re.compile(r"(0\d{1,2})-((\d{3,4})-(\d{4}))")
result = p.finditer(txt)
for tel in result:
    print(tel.group(1))
    print(tel.group(2))
    print(tel.group(3))
    print(tel.group(4))
    print("-----------------")

010
2323-3232
2323
3232
-----------------
02
323-9999
323
9999
-----------------
031
233-4841
233
4841
-----------------
010
1112-3344
1112
3344
-----------------


In [ ]:
txt = "2020/10/13 22:32:23"
# 1((2 )/(3 )/(4 )) 5((6 ):(7 ):(8 ))
result = re.finditer(r"((\d{4})/(\d{2})/(\d{2})) ((\d{2}):(\d{2}):(\d{2}))", txt)
for v in result:
    print(v.group())
    print(v.group(1))
    print(v.group(2))
    print(v.group(3))
    print(v.group(4))
    print(v.group(5))
    print(v.group(6))
    print(v.group(7))
    print(v.group(8))

2020/10/13 22:32:23
2020/10/13
2020
10
13
22:32:23
22
32
23


In [99]:
txt = "2020/10/13 22:32:23"
# 1((2 )/(3 )/(4 )) 5((6 ):(7 ):(8 ))
result = re.findall(r"((\d{4})/(\d{2})/(\d{2})) ((\d{2}):(\d{2}):(\d{2}))", txt)
result


[('2020/10/13', '2020', '10', '13', '22:32:23', '22', '32', '23')]

### 패턴 안에서 하위 패턴 참조 지정
- `\번호`
- 지정한 '번호' 번째 패턴으로 매칭된 문자열과 같은 문자열을 의미

In [106]:
txt = "010-1111-1111, 010-1234-5678, 010-3333-3333, 010-2222-0987"

# 국번과 번호가 같은 전화번호를 조회.
p = re.compile(r"0\d{1,2}-(\d{3,4})-\1")  #\1 -> 1번 하위그룹과 같은 값
result = p.finditer(txt)
for tel in result:
    print(tel)

<re.Match object; span=(0, 13), match='010-1111-1111'>
<re.Match object; span=(30, 43), match='010-3333-3333'>


### 패턴과 매칭된 결과의 일부분만 변경

In [ ]:
info ='''김정수 kjs@gmail.com 801023-1010221   
박영수 pys.abc@gmail.com 700121-1120212
이민영 lmy-abc@naver.com 820301-2020122
김순희 ksh@daum.net 781223-2012212
오주연 ojy@daum.net 900522-1023218
'''

In [113]:
# 주민번호 뒤 6자리를 가리기(masking)
p = re.compile(r"(\d{6}-[1-8])\d{6}") # 나올 부분을 subgroup으로 묶는다.
result = p.sub(r"\g<1>######", info) 
# 변환할 문자열을 만들때 "\g<1>" 1번 subgroup으로 찾은 문자열을 그대로 나오게한다.
print(result)

김정수 kjs@gmail.com 801023-1######   
박영수 pys.abc@gmail.com 700121-1######
이민영 lmy-abc@naver.com 820301-2######
김순희 ksh@daum.net 781223-2######
오주연 ojy@daum.net 900522-1######



In [115]:
# 이름을 masking  # 김순희 -> 김**
p2 = re.compile(r"([가-힣])[가-힣]{2,}")
result2 = p2.sub(r"\g<1>**", result)
print(result2)

김** kjs@gmail.com 801023-1######   
박** pys.abc@gmail.com 700121-1######
이** lmy-abc@naver.com 820301-2######
김** ksh@daum.net 781223-2######
오** ojy@daum.net 900522-1######



### group으로 묶인 것 참조(조회)
- **패턴 안에서 참조**
    - `\번호` , `r'(\d{3}) \1'` => 중복되는 것을 패턴으로 표현할 때.
- **match 조회**
    - match객체.group(번호)
- **sub() 함수에서 대체 문자로 참조**
    - `\g<번호>`

# Greedy 와 Non-Greedy(Lazy) Matching
정규표현식의 수량자(quantifier)는 기본적으로 greedy(탐욕적) 방식으로 동작한다. 이는 가능한 한 많은 문자를 매칭하려고 시도한다는 의미이다. 반면 non-greedy(lazy, 게으른) 방식은 조건을 만족하는 최소한의 문자만 매칭하는 방식이다.
- **Greedy** matching
    - 정규표현식 엔진이 가능한 한 많은 문자를 매칭하려고 시도한다.
    - 패턴에 만족하는 가장 긴 문자열을 찾는다.
    - 정규표현식의 기본 동작 방식.
    - 예
        - 패턴: `<.+>`
        - 문자열: `<div>Hello</div>`
        - 매칭 결과: `<div>Hello</div>`
- **Non-Greedy(Lazy)** matching
    - 최소한의 문자만 매칭하려고 시도합니다
    - 패턴에 만족하는 가장 짧은 문자열을 찾습니다
    - 수량자 뒤에 `?`를 붙여서 사용합니다
      - `*?`
      - `+?`
      - `{m,n}?`
    - 예
      - 패턴: `<.+?>`
      - 문자열: `<div>Hello</div>`
      - 매칭 결과: `<div>`

In [ ]:
tag = "<div><a href='url'>Hello</a></div>"
# p = re.compile(r"<.+>") # greedy(기본방식)
p = re.compile(r"<.+?>") # non-greedy (수량사?)  <[^/].+?>: 닫는태그는 제외
result = p.findall(tag)
print(len(result))
result

2


['<div>', "<a href='url'>"]

In [121]:
p = re.compile(r"<.+>") # greedy(기본방식)
p = re.compile(r"<.+?>") # non-greedy (수량사?)
p.sub("", tag)

'Hello'

# 전방/후방 탐색 (Lookahead/Lookbehind)

- 특정 패턴의 **앞이나 뒤에 있는 조건을 확인**하지만, **그 조건 부분은 실제 매칭 결과에 포함시키지 않는** 기법이다.
- 이 기능을 통해 문자열 내에서 특정 조건이 충족되는지 확인하면서, 실제 매칭에서는 그 조건을 포함하지 않도록 할 수 있다.
- 탐색 방향에 따라 전방 탐색과 후방 탐색으로 구분되며, 각각 긍정형과 부정형이 있다.

## 전방 탐색 (Lookahead)
**매칭될 문자열 뒤에 오는 조건을 확인**한다. 조건은 매칭 결과에 포함되지 않는다.

- **긍정 전방 탐색 (Positive Lookahead)**
  - **구문**: `xxxxx(?=패턴)`
  - **의미**: 괄호 안의 패턴이 뒤에 오는 경우를 찾는다. 매칭(반환)은 xxxxx 부분만 한다.
  - **예시**: 
    - `\d+(?=원)`: 숫자 뒤에 "원"이 오는 패턴을 찾고 숫자만 매칭한다.
      - 문자열: `1000원, 2000달러` → 매칭: `1000`
    - `[A-Z][a-z]+(?=\s+is)`: " is" 앞의 단어를 매칭한다.
      - 문자열: `John is here` → 매칭: `John`

- **부정 전방 탐색 (Negative Lookahead)**
  - **구문**: `xxxxx(?!패턴)`
  - **의미**: 괄호 안의 패턴이 뒤에 오지 않는 경우를 찾는다. 매칭(반환)은 xxxxx 부분만 한다.
  - **예시**: 
    - `\d+(?!원)`: 숫자 뒤에 "원"이 오지 않는 패턴을 찾고 숫자만 매칭한다.
      - 문자열: `1000원, 2000달러` → 매칭: `2000`
    - `\w+(?!\.txt)`: ".txt"로 끝나지 않는 단어를 매칭한다.
      - 문자열: `file.txt readme.md` → 매칭: `file`, `readme`, `md`

## 후방 탐색 (Lookbehind)
**매칭될 문자열 앞에 오는 조건을 확인**한다. 조건은 매칭 결과에 포함되지 않는다.

- **긍정 후방 탐색 (Positive Lookbehind)**
  - **구문**: `(?<=패턴)xxxxx`
  - **의미**: 괄호 안의 패턴이 앞에 오는 경우를 찾는다. 매칭(반환)은 xxxxx 부분만 한다.
  - **예시**: 
    - `(?<=\$)\d+`: "$" 뒤의 숫자를 매칭한다.
      - 문자열: `$1000, 2000원` → 매칭: `1000`
    - `(?<=@)\w+`: "@" 뒤의 단어를 매칭한다.
      - 문자열: `user@example.com` → 매칭: `example`

- **부정 후방 탐색 (Negative Lookbehind)**
  - **구문**: `(?<!패턴)xxxxx`
  - **의미**: 괄호 안의 패턴이 앞에 오지 않는 경우를 찾는다. 매칭(반환)은 xxxxx 부분만 한다.
  - **예시**: 
    - `(?<!\$)\d+`: "$"가 앞에 없는 숫자를 매칭한다.
      - 문자열: `$1000, 2000원` → 매칭: `2000`
    - `(?<!un)\w+able`: "un"으로 시작하지 않는 "-able" 단어를 매칭한다.
      - 문자열: `comfortable, unable` → 매칭: `comfortable`

In [ ]:
info1 = """TV 30000원 30개
컴퓨터 32000원 50개
모니터 15000원 70개"""

p = re.compile(r"\d+(?=원)")
p.findall(info1)

['30000', '30', '32000', '50', '15000', '70']

In [127]:
info2 = """TV $30000 30개
컴퓨터 $32000 50개
모니터 $15000 70개"""

p2 = re.compile(r"(?<=\$)\d+")
p2.findall(info2)

['30000', '32000', '15000']